### PR TITLE
Client APIs for downloading/uploading trace data

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/11456?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11456/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11456
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Client APIs for downloading/uploading trace data.

Title

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

```python
import mlflow
from mlflow.tracing.types.model import TraceData


mlflow.set_tracking_uri("databricks")
name = "/Users/harutaka.kawamura@databricks.com/test"
mlflow.set_experiment(name)
exp = mlflow.get_experiment_by_name(name)

client = mlflow.MlflowClient()
trace_info = client.create_trace(
    experiment_id=exp.experiment_id,
    timestamp_ms=123,
    execution_time_ms=567,
    status="OK",
)


data = TraceData(spans=[])
client._tracking_client._upload_trace_data(trace_info.request_id, data)
downloaded = client._tracking_client._download_trace_data(trace_info.request_id)
print(downloaded)
```

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
@mlflow-automation autoformat

**Comment by harupy** [comment]
@mlflow-automation autoformat

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_artifact_repo.py:141]: This is to avoid fetching a run and hitting a `NotFound` error.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_artifact_repo.py:164]: The endpoints for trace upload and download look like `"/mlflow/traces/{trace_id}/credentials-for-data-{download|upload}"`. `trace_id` needs to be replaced.

File: mlflow/tracking/_tracking_service/client.py
Comment (harupy) [inline mlflow/tracking/_tracking_service/client.py:625]:
```suggestion
```

just for testing

File: mlflow/tracking/client.py
Comment (harupy) [inline mlflow/tracking/client.py:447]:
```suggestion
```

just for testing

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (BenWilson2) [reply mlflow/store/artifact/databricks_artifact_repo.py:142]: The previous behavior set the `self.run_relative_artifact_repo_root_path` to `""` at init time. 
This is a behavior change, right?

File: mlflow/tracking/_tracking_service/client.py
Comment (dbczumar) [inline mlflow/tracking/_tracking_service/client.py:576]: @harupy if you don't mind, can we update the backend so that attribute is called `mlflow.artifactLocation` and update the client too?

File: mlflow/tracking/client.py
Comment (dbczumar) [inline mlflow/tracking/client.py:446]: Reminder to remove :)

File: mlflow/tracking/_tracking_service/client.py
Comment (dbczumar) [inline mlflow/tracking/_tracking_service/client.py:620]: This method is only defined on `DatabricksArtifactRepository`. Can we add `download_trace` and `upload_trace` to the base artifact repository class and raise `NotImplemented` by default? Can we also file a follow-up ticket to track implementing `download_trace` and `upload_trace` for all of our other artifact repos?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:125]: Can we add docstrings explaining what these properties are and when they may or may not be defined & with what values?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [reply mlflow/store/artifact/databricks_artifact_repo.py:142]: Now we're initializing this value as soon as it's used, as opposed to during `__init__`. Seems fairly reasonable to me.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:161]: Can we add a docstring to this method while we're here, and can we document the parameters?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:262]: Shall we make `json_body` optional so we don't have to pass `None`?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:263]: Nit: Can we specify the kwarg names here?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:235]: Nit: Can we specify the kwarg names here?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:221]: Nit: Can we add type hints to `download_trace` and `upload_trace`?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:247]: what happens if it can't be parsed as JSON? Can we throw an informative exception and add a test case?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:283]: Should we be reusing `_azure_adls_gen2_upload_file` instead? One thing that's missing from the current impl is retries. `_azure_adls_gen2_upload_file` uses `_retryable_adls_function`

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:48]: Is this used?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:249]: We should only be uploading the `TraceData`, not the full `Trace`

Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:249]: Also, is it a bit strange that `upload_trace` takes a `Trace` (should be `TraceData`) object, but `download_trace` returns a Python dictionary? IMO I think they should both take a dict. The client layer should be responsible for validating / transforming the TraceData, not the artifact repo.

File: mlflow/tracking/_tracking_service/client.py
Comment (harupy) [reply mlflow/tracking/_tracking_service/client.py:620]: Got it, filed a ticket!

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (harupy) [reply mlflow/store/artifact/databricks_artifact_repo.py:254]: do we need two arguments? trace_id and trace_data?

Comment (dbczumar) [reply mlflow/store/artifact/databricks_artifact_repo.py]: Can we get trace ID from the artifact URI?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (harupy) [reply mlflow/store/artifact/databricks_artifact_repo.py:286]: Do we want to retry? `_signed_url_upload_file` doesn't do a retry.

Comment (harupy) [reply mlflow/store/artifact/databricks_artifact_repo.py:286]: `_retryable_adls_function` uses `self._get_write_credential_infos` to generate a new credential. That needs to be changed.

Comment (harupy) [reply mlflow/store/artifact/databricks_artifact_repo.py:286]: For ADLS, we always use multipart-upload. We don't want to retry the upload just because one chunk upload fails. I think that's the reason.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (harupy) [reply mlflow/store/artifact/databricks_artifact_repo.py]: Updated the function signature

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_artifact_repo.py:126]: This can be removed.

File: mlflow/tracking/client.py
Comment (harupy) [inline mlflow/tracking/client.py:446]:
```suggestion
```

**Comment by harupy** [comment]
I'll add unit tests.

File: mlflow/tracking/_tracking_service/client.py
Comment (harupy) [inline mlflow/tracking/_tracking_service/client.py:628]: Should this be kept?

Comment (dbczumar) [reply mlflow/tracking/_tracking_service/client.py:628]: The request ID shouldn't be in the API signature for `upload_trace()` or `download_trace()`, right? We should be able to identify the request ID by parsing the artifact URI

Comment (dbczumar) [reply mlflow/tracking/_tracking_service/client.py:628]: Sorry, ignore my last comment if you saw it. No, we shouldn't keep this.

File: mlflow/tracking/_tracking_service/client.py
Comment (dbczumar) [inline mlflow/tracking/_tracking_service/client.py:634]: We can remove this too.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [reply mlflow/store/artifact/databricks_artifact_repo.py:278]: Makes sense. As discussed offline, we can add retries later if necessary

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/databricks_artifact_repo.py:166]: Can you give an example?

File: mlflow/store/artifact/artifact_repo.py
Comment (dbczumar) [inline mlflow/store/artifact/artifact_repo.py:336]:
```suggestion
    def download_trace_data(self) -> Dict[str, Any]:
        """
        Download the trace file for the given trace_id.

        Returns:
            The trace data as a dictionary.
        """
        raise NotImplementedError

    def upload_trace_data(self, trace_data: Dict[str, Any]) -> None:
```

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (harupy) [reply mlflow/store/artifact/databricks_artifact_repo.py:166]: Example of `path_params`?

File: mlflow/protos/databricks_artifacts.proto
Comment (harupy) [inline mlflow/protos/databricks_artifacts.proto:83]: test

File: mlflow/tracing/types/model.py
Comment (harupy) [inline mlflow/tracing/types/model.py:79]: These need to be implemented in other data classes.

File: mlflow/tracking/client.py
Comment (dbczumar) [inline mlflow/tracking/client.py:446]: Let's remove before merge :)

File: tests/store/artifact/test_databricks_artifact_repo.py
Comment (dbczumar) [inline tests/store/artifact/test_databricks_artifact_repo.py:1515]: Can we verify that a real tracedata can be uploaded / downloaded, instead of a fake dict?

File: tests/store/artifact/test_databricks_artifact_repo.py
Comment (dbczumar) [inline tests/store/artifact/test_databricks_artifact_repo.py:1525]: This method name seems to be outdated

File: tests/store/artifact/test_databricks_artifact_repo.py
Comment (dbczumar) [inline tests/store/artifact/test_databricks_artifact_repo.py:1545]: This method name appears to be outdated

**Review by dbczumar — APPROVED** [review]
LGTM once comments are addressed :)

File: mlflow/protos/databricks_artifacts.proto
Comment (dbczumar) [inline mlflow/protos/databricks_artifacts.proto:103]: Can we rename these in accordance with the backend renaming?

File: tests/store/artifact/test_databricks_artifact_repo.py
Comment (harupy) [reply tests/store/artifact/test_databricks_artifact_repo.py:1545]: thanks for the catch, updated!

File: mlflow/protos/databricks_artifacts.proto
Comment (harupy) [reply mlflow/protos/databricks_artifacts.proto:103]: updated!

File: tests/store/artifact/test_databricks_artifact_repo.py
Comment (harupy) [reply tests/store/artifact/test_databricks_artifact_repo.py]: Updated!

**Pull request merged**
